Module.php

```diff
@@ -258,6 +440,81 @@ protected function listFiles(string $directory, bool $recursive = false): array
         return array_values(array_unique(array_merge($listRootFiles, $listFiles)));
     }
 
+    /**
+     * Get all directories available to sideload.
+     */
+    protected function listDirs(string $directory, int $maxDepth = -1, int $maxDirectories = 0): array
+    {
+        $listDirs = [];
+
+        $dir = new \SplFileInfo($directory);
+        if (!$dir->isDir()) {
+            return [];
+        }
+
+        $countDirs = 0;
+        $this->directory = $directory;
+
+        $lengthDir = strlen($directory) + 1;
+        $dir = new \RecursiveDirectoryIterator($directory);
+        // Prevent UnexpectedValueException "Permission denied" by excluding
+        // directories that are not executable or readable.
+        $dir = new \RecursiveCallbackFilterIterator($dir, function ($current, $key, $iterator) {
+            if ($iterator->isDir() && (!$iterator->isExecutable() || !$iterator->isReadable())) {
+                return false;
+            }
+            return true;
+        });
+        // Follow the same rules than SideloadDir::listDirs, even if empty dirs
+        // may be allowed here.
+        $iterator = new \RecursiveIteratorIterator($dir);
+        $iterator->setMaxDepth($maxDepth);
+        /** @var \SplFileInfo $file */
+        foreach ($iterator as $filepath => $file) {
+            if ($file->isDir()) {
+                if ($this->verifyFileOrDir($file, true)) {
+                    // There are two filepaths for one dirpath: "." and "..".
+                    $filepath = $file->getRealPath();
+                    // Don't list empty directories.
+                    if (!$this->dirHasNoFileAndIsRemovable($filepath)) {
+                        // For security, don't display the full path to the user.
+                        $relativePath = substr($filepath, $lengthDir);
+                        if (!isset($listDirs[$relativePath])) {
+                            // Use keys for quicker process on big directories.
+                            $listDirs[$relativePath] = null;
+                            if ($maxDirectories && ++$countDirs >= $maxDirectories) {
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        $listDirs = array_keys($listDirs);
+        natcasesort($listDirs);
+        return array_combine($listDirs, $listDirs);
+    }
+
+    /**
+     * Check if a directory, that is valid, contains files or unwriteable content, recursively.
+     *
+     * The directory should be already checked.
+     */
+    private function dirHasNoFileAndIsRemovable(string $dir): bool
```

Comment:
As with listDirs(), I see that there are two dirHasNoFileAndIsRemovable() methods that are more-or-less identical. I prefer to minimize code duplication.